jscomp/syntax/reactjs_jsx_ppx_v3.ml

```diff
@@ -0,0 +1,1092 @@
+      end
+      | structures -> begin
+        default_mapper.structure mapper @@ reactComponentTransform mapper structures
+      end
+    ) in
+
+  let expr =
+    (fun mapper expression -> match expression with
+       (* Does the function application have the @JSX attribute? *)
+       | {
+           pexp_desc = Pexp_apply (callExpression, callArguments);
+           pexp_attributes
+         } ->
+         let (jsxAttribute, nonJSXAttributes) = List.partition (fun (attribute, _) -> attribute.txt = "JSX") pexp_attributes in
+         (match (jsxAttribute, nonJSXAttributes) with
+         (* no JSX attribute *)
+         | ([], _) -> default_mapper.expr mapper expression
+         | (_, nonJSXAttributes) -> transformJsxCall mapper callExpression callArguments nonJSXAttributes)
+
+       (* is it a list with jsx attribute? Reason <>foo</> desugars to [@JSX][foo]*)
+       | {
+           pexp_desc =
+            Pexp_construct ({txt = Lident "::"; loc}, Some {pexp_desc = Pexp_tuple _})
+            | Pexp_construct ({txt = Lident "[]"; loc}, None);
+           pexp_attributes
+         } as listItems ->
+          let (jsxAttribute, nonJSXAttributes) = List.partition (fun (attribute, _) -> attribute.txt = "JSX") pexp_attributes in
+          (match (jsxAttribute, nonJSXAttributes) with
+          (* no JSX attribute *)
+          | ([], _) -> default_mapper.expr mapper expression
+          | (_, nonJSXAttributes) ->
+            let fragment = Exp.ident ~loc {loc; txt = Ldot (Lident "ReasonReact", "fragment")} in
+            let childrenExpr = transformChildrenIfList ~loc ~mapper listItems in
+            let args = [
+              (* "div" *)
+              (nolabel, fragment);
+              (* [|moreCreateElementCallsHere|] *)
+              (nolabel, childrenExpr)
+            ] in
+            Exp.apply
+              ~loc
+              (* throw away the [@JSX] attribute and keep the others, if any *)
+              ~attrs:nonJSXAttributes
+              (* ReactDOMRe.createElement *)
+              (Exp.ident ~loc {loc; txt = Ldot (Lident "ReactDOMRe", "createElement")})
+              args
+         )
+       (* Delegate to the default mapper, a deep identity traversal *)
+       | e -> default_mapper.expr mapper e) in
+
+  let module_binding =
+    (fun mapper module_binding ->
+      let _ = nestedModules := module_binding.pmb_name.txt :: !nestedModules in
+      let mapped = default_mapper.module_binding mapper module_binding in
+      let _ = nestedModules := List.tl !nestedModules in
+      mapped
+    ) in
+
+  { default_mapper with structure; expr; signature; module_binding; }
+
+
+#if BS_COMPILER_IN_BROWSER then
+(* keep in sync with jscomp/core/jsoo_main.ml `let implementation` *)
+let rewrite (code: Parsetree.structure) =
+  let mapper = jsxMapper () in
+  Location.input_name := "Toplevel";
+  mapper.structure mapper code
+#else
+let () = Ast_mapper.register "JSX" (fun _argv -> jsxMapper ())
```

Comment:
`node ./scripts/ninja.js && ninja -C jscomp` fails when this is written this way, even though this is "correct", but repl.js seems to set BS_COMPILER_IN_BROWSER and work properly. I think the env isn't set properly in ninja -C jscomp - any ideas?